### PR TITLE
feat: introduce `autoscaler_policy_definitions` to  `autoscaler_policies`

### PR DESCRIPTION
**Pull request by aweris**
Introduces `autoscaler_policy_definitions` to `autoscaler_policies` to allow migration of the old JSON field `autoscaler_policies_json` to the proper Terraform configuration.

current policy definition:

```hcl
resource "castai_autoscaler" "castai_autoscaler_policies" {
  cluster_id = "cluster-id"

  autoscaler_policies_json = <<-EOT
     {
        "enabled": true,
        "unschedulablePods": {
            "enabled": true
        },
        "nodeDownscaler": {
            "enabled": true,
            "emptyNodes": {
                "enabled": true
            },
            "evictor": {
                "aggressiveMode": false,
                "cycleInterval": "5m10s",
                "dryRun": false,
                "enabled": true,
                "nodeGracePeriodMinutes": 10,
                "scopedMode": false
            }
        },
        "nodeTemplatesPartialMatchingEnabled": false,
        "clusterLimits": {
            "cpu": {
                "maxCores": 20,
                "minCores": 1
            },
            "enabled": true
        }
    }
  EOT
}
```

new resource definition:
```hcl
resource "castai_autoscaler" "castai_autoscaler_policies" {
  cluster_id = "b28629f0-485f-4a29-bbc0-0e3feb3dbbfa"

  autoscaler_policy_definitions {
    enabled = true
    node_templates_partial_matching_enabled = false

    unschedulable_pods {
      enabled = true
    }

    node_downscaler {
      enabled = true

      empty_nodes {
        enabled = false
      }

      evictor {
        aggressive_mode           = false
        cycle_interval            = "5m10s"
        dry_run                   = false
        enabled                   = true
        node_grace_period_minutes = 10
        scoped_mode               = false
      }
    }

    cluster_limits {
      cpu {
        max_cores = 20
        min_cores = 1
      }
      enabled = true
    }
  }
}
```

TODO: 

- [x] update docs

Depends on #321

### CODE REVIEW

File: castai/resource_autoscaler.go [hunk 1]
```diff
@@ -14,13 +14,53 @@ import (
...
+	FieldMinCores                                 = "min_cores"
+	FieldMaxCores                                 = "max_cores"
+	FieldSpotInstances                            = "spot_instances"
+	FieldMaxReclaimRate                           = "max_reclaim_rate"
+	FieldSpotBackups                              = "spot_backups"
+	FieldSpotDiversityEnabled                     = "spot_diversity_enabled"
+	FieldSpotDiversityPriceIncreaseLimit          = "spot_diversity_price_increase_limit_percent"
+	FieldSpotInterruptionPredictions              = "spot_interruption_predictions"
+	FieldSpotBackupRestoreRateSeconds             = "spot_backup_restore_rate_seconds"
+	FieldSpotInterruptionPredictionsType          = "type"
```
[the remark this answers is not in the document]
Comment (aweris) [reply castai/resource_autoscaler.go:52]: `spot_interruption_predictions_type` is typo, missed it. However dry_run is correct according to our json struct

File: castai/resource_autoscaler.go [hunk 2]
```diff
@@ -49,12 +89,381 @@ func resourceAutoscaler() *schema.Resource {
...
+												FieldMinCPUCores: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     0,
+													Description: "defines min CPU cores for the node to pick.",
+												},
+												FieldMaxCPUCores: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     32,
```
[the remark this answers is not in the document]
Comment (aweris) [reply castai/resource_autoscaler.go:214]: this is the default value provided by the Proto file. If `FieldNodeConstraints` is empty, we're not going to set anything. However, if the user sets `FieldNodeConstraints` partially, this is the default value we're going to use.

File: castai/resource_autoscaler.go [hunk 3]
```diff
@@ -49,12 +89,381 @@ func resourceAutoscaler() *schema.Resource {
...
+										Description: "defines the minimum and maximum amount of CPUs for cluster's worker nodes.",
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												FieldMinCores: {
+													Type:             schema.TypeInt,
+													Optional:         true,
+													Default:          1,
+													Description:      "defines the minimum allowed amount of CPUs in the whole cluster.",
+													ValidateDiagFunc: validation.ToDiagFunc(validation.IntBetween(1, 20)),
+												},
```
Comment (linkas45) [inline castai/resource_autoscaler.go:273]: why only 1 to 20 when max can be 2 to 1000?

Comment (aweris) [reply castai/resource_autoscaler.go:273]: These values are based on our api contract definitons

```proto
// Defines the minimum and maximum amount of vCPUs for cluster's worker nodes.
message ClusterLimitsCpu {
  // Defines the minimum allowed amount of CPUs in the whole cluster.
  int32 min_cores = 1 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = { default: "1", minimum: 1, maximum: 20, example: "1" }];
  // Defines the maximum allowed amount of vCPUs in the whole cluster.
  int32 max_cores = 2 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = { default: "20", minimum: 2, maximum: 1000, example: "20" }];
}
```

https://gitlab.com/castai/kubecast/-/blob/master/api-contracts/grpc/proto/policies/v1/policy.proto?ref_type=heads#L110

File: castai/resource_autoscaler.go [hunk 4]
```diff
@@ -49,12 +89,381 @@ func resourceAutoscaler() *schema.Resource {
 				Description:      "autoscaler policies JSON string to override current autoscaler settings",
 				Optional:         true,
 				ValidateDiagFunc: validateAutoscalerPolicyJSON(),
+				Deprecated:       "use autoscaler_policy_overrides instead",
```
Comment (linkas45) [inline castai/resource_autoscaler.go:92]: can you update this message to include link to readme for examples?

**Review by linkas45 — APPROVED** [review]

File: castai/resource_autoscaler.go (hunk 4 above)
Comment (aweris) [reply castai/resource_autoscaler.go:92]: is it necessary, long messages will be reduce readability of this since it's part of the terraform plan

Comment (linkas45) [reply castai/resource_autoscaler.go:92]: I think it's a good practise to inform customer regarding next steps. I think we usually put these links we such changes.

Comment (aweris) [reply castai/resource_autoscaler.go:92]: Sure, which link are we going to provide? readme of github or Terraform resource documentation?

Comment (linkas45) [reply castai/resource_autoscaler.go:92]: readme of github, the one that contains example

Comment (aweris) [reply castai/resource_autoscaler.go:92]: Done: 
```hcl
Plan: 0 to add, 1 to change, 0 to destroy.
╷
│ Warning: Argument is deprecated
│ 
│   with castai_autoscaler.castai_autoscaler_policies,
│   on castai.tf line 23, in resource "castai_autoscaler" "castai_autoscaler_policies":
│   23:   autoscaler_policies_json = "{}"
│ 
│ use autoscaler_policy_overrides instead. See README for examples: https://github.com/castai/terraform-provider-castai?tab=readme-ov-file#migrating-from-6xx-to-7xx
│ 
│ (and one more similar warning elsewhere)
```

Comment (linkas45) [reply castai/resource_autoscaler.go]: nice

**Pull request merged**